src/docfx/build/document/RepositoryProvider.cs

```diff
@@ -27,70 +29,93 @@ internal class RepositoryProvider
         {
-            switch (origin)
+            return origin switch
+            {
+                FileOrigin.Default => (_docsetPath, DefaultRepository),
+                FileOrigin.Fallback when _localizationProvider != null
+                    => _localizationProvider.GetFallbackRepositoryWithDocsetEntry(),
+                FileOrigin.Dependency when _config != null && _packageResolver != null && dependencyName != null
+                    => _dependencyRepositories.GetOrAdd(dependencyName.Value, key => GetDependencyRepository(key, _config, _packageResolver)),
+                _ => throw new InvalidOperationException(),
+            };
+        }
+
+        private (Repository? repository, PathString? pathToRepository) GetRepository(string fullPath)
+        {
+            var directory = Path.GetDirectoryName(fullPath);
+            if (directory is null)
+            {
+                return default;
+            }
+
+            var repository = _repositores.GetOrAdd(directory, GetRepositoryCore);
+            if (repository is null)
             {
-                case FileOrigin.Redirection:
-                case FileOrigin.Default:
-                    return (_docsetPath, _repository);
-
-                case FileOrigin.Fallback when _localizationProvider != null:
-                    return _localizationProvider.GetFallbackRepositoryWithDocsetEntry();
-
-                case FileOrigin.Template when _config != null && _packageResolver != null:
-                    return _templateRepository.Value;
-
-                case FileOrigin.Dependency when _config != null && _packageResolver != null && dependencyName != null:
-                    return _dependencyRepositories.GetOrAdd(dependencyName.Value, _ => new Lazy<(string docset, Repository? repository)>(() =>
-                    {
-                        var dependency = _config.Dependencies[dependencyName.Value];
-                        var dependencyPath = _packageResolver.ResolvePackage(dependency, dependency.PackageFetchOptions);
-
-                        if (dependency.Type != PackageType.Git)
-                        {
-                            // point to a folder
-                            return (dependencyPath, null);
-                        }
-
-                        return (dependencyPath, Repository.Create(dependencyPath, dependency.Branch, dependency.Url));
-                    })).Value;
+                return default;
             }
 
-            throw new InvalidOperationException();
+            return (repository, new PathString(Path.GetRelativePath(repository.Path, fullPath)));
         }
 
-        private (string path, Repository?) GetTemplateRepository()
+        private Repository? GetRepositoryCore(string directory)
         {
-            if (_config is null || _packageResolver is null)
+            var repository = GitUtility.FindRepository(directory);
```

Comment:
consider rename `repository` to `repoPath`

Reply:
make sense